File: src/main/java/com/gg/plugins/mongo/utils/GuiUtils.java
/*
 * Copyright (c) 2018 David Boissier.
 * Modifications Copyright (c) 2022 Geetesh Gupta.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.gg.plugins.mongo.utils;

import com.intellij.openapi.actionSystem.ActionManager;
import com.intellij.openapi.actionSystem.ActionToolbar;
import com.intellij.openapi.actionSystem.DefaultActionGroup;
import com.intellij.openapi.ui.MessageType;
import com.intellij.openapi.ui.popup.Balloon;
import com.intellij.openapi.ui.popup.JBPopupFactory;
import com.intellij.openapi.util.IconLoader;
import com.intellij.ui.awt.RelativePoint;
import com.intellij.util.ui.UIUtil;

import javax.swing.*;
import java.awt.*;

public class GuiUtils {

	private static final String ICON_FOLDER = "/icons/";

	public static Icon loadIcon(String iconFilename) {
		return IconLoader.findIcon(ICON_FOLDER + iconFilename);
	}

	public static Icon loadIcon(String iconFilename, String darkIconFilename) {
		String iconPath = ICON_FOLDER;
		if (isUnderDarcula()) {
			iconPath += darkIconFilename;
		} else {
			iconPath += iconFilename;
		}
		return IconLoader.findIcon(iconPath);
	}

	private static boolean isUnderDarcula() {
		return UIManager.getLookAndFeel().getName().contains("Darcula");
	}

	public static void installActionGroupInToolBar(DefaultActionGroup actionGroup,
			JPanel toolBarPanel,
			ActionManager actionManager,
			String toolbarName,
			boolean horizontal) {
		if (actionManager == null) {
			return;
		}

		ActionToolbar a = ActionManager.getInstance().createActionToolbar(toolbarName, actionGroup, horizontal);
		a.setTargetComponent(a.getComponent());
		JComponent actionToolbar = a.getComponent();
		toolBarPanel.add(actionToolbar, BorderLayout.CENTER);
	}

	public static Dimension enlargeWidth(Dimension preferredSize, double factor) {
		int enlargedWidth = Double.valueOf(preferredSize.width * factor).intValue();
		return new Dimension(enlargedWidth, preferredSize.height);
	}

	public static void showNotification(final JComponent component,
			final MessageType info,
			final String message,
			final Balloon.Position position) {
		UIUtil.invokeLaterIfNeeded(() -> JBPopupFactory.getInstance()
		                                               .createBalloonBuilder(new JLabel(message))
		                                               .setFillColor(info.getPopupBackground())
		                                               .createBalloon()
		                                               .show(new RelativePoint(component, new Point(0, 0)), position));
	}
}

File: src/main/kotlin/com/gg/plugins/mongo/utils/GuiUtils.kt
/*
 * Copyright (c) 2018 David Boissier.
 * Modifications Copyright (c) 2022 Geetesh Gupta.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.gg.plugins.mongo.utils

import com.intellij.openapi.actionSystem.ActionManager
import com.intellij.openapi.actionSystem.DefaultActionGroup
import com.intellij.openapi.ui.MessageType
import com.intellij.openapi.ui.popup.Balloon
import com.intellij.openapi.ui.popup.JBPopupFactory
import com.intellij.openapi.util.IconLoader
import com.intellij.ui.awt.RelativePoint
import com.intellij.util.ui.UIUtil
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.Point
import javax.swing.Icon
import javax.swing.JComponent
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.UIManager

object GuiUtils {
    private const val ICON_FOLDER = "/icons/"
    fun loadIcon(iconFilename: String): Icon? {
        return IconLoader.findIcon(ICON_FOLDER + iconFilename, GuiUtils::class.java)
    }

    fun loadIcon(iconFilename: String?, darkIconFilename: String?): Icon? {
        var iconPath: String? = ICON_FOLDER
        iconPath += if (isUnderDarcula) {
            darkIconFilename
        } else {
            iconFilename
        }
        return IconLoader.findIcon(iconPath!!, GuiUtils::class.java)
    }

    private val isUnderDarcula: Boolean
        get() = UIManager.getLookAndFeel().name.contains("Darcula")

    fun installActionGroupInToolBar(
        actionGroup: DefaultActionGroup?,
        toolBarPanel: JPanel?,
        actionManager: ActionManager?,
        toolbarName: String?,
        horizontal: Boolean
    ) {
        if (actionManager == null) {
            return
        }
        val a = ActionManager.getInstance().createActionToolbar(toolbarName!!, actionGroup!!, horizontal)
        a.targetComponent = a.component
        val actionToolbar = a.component
        toolBarPanel!!.add(actionToolbar, BorderLayout.CENTER)
    }

    fun enlargeWidth(preferredSize: Dimension, factor: Double): Dimension {
        val enlargedWidth = java.lang.Double.valueOf(preferredSize.width * factor).toInt()
        return Dimension(enlargedWidth, preferredSize.height)
    }

    fun showNotification(
        component: JComponent?,
        info: MessageType,
        message: String?,
        position: Balloon.Position?
    ) {
        UIUtil.invokeLaterIfNeeded {
            JBPopupFactory.getInstance()
                .createBalloonBuilder(JLabel(message))
                .setFillColor(info.popupBackground)
                .createBalloon()
                .show(RelativePoint(component!!, Point(0, 0)), position)
        }
    }
}